src/Lean/Linter/Basic.lean

```diff
@@ -5,18 +5,71 @@ Authors: Lars König
 -/
 prelude
 import Lean.Data.Options
+import Lean.MonadEnv
 import Lean.Log
 
 namespace Lean.Linter
 
+/-- Linter sets are represented as a map from linter name to set name,
+to make it easy to look up which sets to check for enabling a linter.
+-/
+def LinterSets := NameMap (Array Name)
+  deriving EmptyCollection, Inhabited
+
+/-- Insert a set into a `LinterSets` map.
+
+`entry.1` is the name of the linter set,
+`entry.2` contains the names of the set's linter options.
+-/
+def insertLinterSetEntry (map : LinterSets) (entry : Name × Array Name) : LinterSets :=
+  entry.2.foldl (init := map) fun map linterName =>
+    map.insert linterName ((map.findD linterName #[]).push entry.1)
+
+builtin_initialize linterSetsExt : SimplePersistentEnvExtension (Name × Array Name) LinterSets ← Lean.registerSimplePersistentEnvExtension {
+  addImportedFn := mkStateFromImportedEntries insertLinterSetEntry {}
+  addEntryFn := insertLinterSetEntry
+  toArrayFn es := es.toArray
+}
+
+structure LinterOptions where
+  toOptions : Options
+  linterSets : LinterSets
+
+def LinterOptions.get [KVMap.Value α] (o : LinterOptions) := o.toOptions.get (α := α)
+def LinterOptions.get? [KVMap.Value α] (o : LinterOptions) := o.toOptions.get? (α := α)
+
+def _root_.Lean.Options.toLinterOptions [Monad m] [MonadEnv m] (o : Options) : m LinterOptions := do
+  let linterSets := linterSetsExt.getState (← getEnv)
+  return { toOptions := o, linterSets }
+
+/-- Return the set of linter sets that this option is contained in. -/
+def LinterOptions.getSet (o : LinterOptions) (opt : Lean.Option α) : Array Name :=
+  o.linterSets.findD opt.name #[]
+
+class MonadLinterOptions (m : Type → Type) where
+  getLinterOptions : m LinterOptions
+
+export MonadLinterOptions (getLinterOptions)
+
+instance [Monad m] [MonadLinterOptions m] : MonadOptions m where
+  getOptions := do return (← getLinterOptions).toOptions
```

Comment:
Why would you ever want this instance?

Reply:
This is needed to call `logLint` (taking a `MonadOptions`) from `logLintIf` (taking a `MonadLinterOptions`).